### PR TITLE
HIP-RA-X technical review; Migrate redundant GEOPHIRES water property functions to utils IAPWS implementations

### PR DESCRIPTION
**Pull request by softwareengineerprogrammer**
1. https://github.com/softwareengineerprogrammer/GEOPHIRES-X/pull/7: https://github.com/NREL/GEOPHIRES-X/issues/109

1. https://github.com/NREL/GEOPHIRES-X/issues/101

* Moved redundant water property calculations in `Reservoir` and `WellBores` to call IAPWS implementations
* In some cases, existing examples generate temperatures beyond IAPWS bounds, added previous calculations as fallbacks for a workaround, marked with `FIXME TODO` to address as part of wider SHR work
* Most example results were impacted by a small percentage by the updated water property calculations, hence the large diff.

1. https://github.com/NREL/GEOPHIRES-X/issues/95
1. https://github.com/NREL/GEOPHIRES-X/issues/105

---

Changes covered by updated unit tests, also updated & tested web interface.

### CODE REVIEW

File: src/geophires_x/Reservoir.py [hunk 1]
```diff
@@ -8,54 +8,8 @@
...
-    Twater = (Twater + 273.15) / 1000
-    A = -203.6060
-    B = 1523.290
-    C = -3196.413
-    D = 2474.455
-    E = 3.855326
-    # water specific heat capacity in J/kg-K
-    cpwater = (A + B * Twater + C * Twater ** 2 + D * Twater ** 3 + E / (Twater ** 2)) / 18.02 * 1000
-    return cpwater
-
```
Comment (softwareengineerprogrammer) [inline src/geophires_x/Reservoir.py:58]: Calculations moved to GeoPHIRESUtils fallback calcuations

File: src/geophires_x/GeoPHIRESUtils.py [hunk 2]
```diff
@@ -82,7 +81,7 @@
 
 
 @lru_cache(maxsize=None)
-def DensityWater(Twater_degC: float) -> float:
+def DensityWater(Twater_degC: float, enable_fallback_calculation=False) -> float:
```
Comment (softwareengineerprogrammer) [inline src/geophires_x/GeoPHIRESUtils.py:84]: Previous calculations moved to fallback calculation in try/catch

File: src/geophires_x/GeoPHIRESUtils.py [hunk 3]
```diff
@@ -94,11 +93,17 @@ def DensityWater(Twater_degC: float) -> float:
...
+        raise ValueError(f'Twater_degC ({Twater_degC}) must be a float or convertible to float.')
 
     try:
         return IAPWS97(T=celsius_to_kelvin(Twater_degC), x=0).rho
     except NotImplementedError as nie:
+        if enable_fallback_calculation:
+            Twater_K = celsius_to_kelvin(Twater_degC)
+            # water density correlation as used in Geophires v1.2 [kg/m3]
+            rho_water = (.7983223 + (1.50896E-3 - 2.9104E-6 * Twater_K) * Twater_K) * 1E3
+            return rho_water
```
Comment (softwareengineerprogrammer) [inline src/geophires_x/GeoPHIRESUtils.py:105]: Previous calculations moved to fallback calculation in try/catch

File: src/hip_ra_x/hip_ra_x.py [hunk 4]
```diff
@@ -186,7 +190,7 @@ def parameter_dict_entry(param: Parameter) -> Parameter:
         )
         self.fluid_heat_capacity: Parameter = parameter_dict_entry(
             floatParameter(
-                'Fluid Heat Capacity',
```
Comment (softwareengineerprogrammer) [inline src/hip_ra_x/hip_ra_x.py:189]: See https://github.com/softwareengineerprogrammer/GEOPHIRES-X/pull/7/files for diff scoped to HIP-RA-X

**Review by malcolm-dsider — APPROVED** [review]
All looks good.

It looks like you are switching to a new code model for printing error message:
 
msg = 'Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....'
model.logger.fatal(msg)
print(msg)

Instead of:
model.logger.fatal('Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....')
print('Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....')

Is it your guidance to do this going forward?

**Comment by softwareengineerprogrammer** [comment]
Thanks for the approval @malcolm-dsider!

> It looks like you are switching to a new code model for printing error message:
> 
> msg = 'Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....' model.logger.fatal(msg) print(msg)
> 
> Instead of: model.logger.fatal('Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....') print('Error: GEOPHIRES failed to Failed to validate CLGS input value. Exiting....')
> 
> Is it your guidance to do this going forward?

My guidance would be to not both print and log - ideally log messages should go to the logger only. However in this case I preserved the existing behavior in case there is some functional reason why both printing and logging were configured - I'm guessing it has to do with incomplete/misunderstood logger configuration that caused someone to add print as a fallback when they weren't seeing messages they expected from the logger. The change I made here is just to define the logged/printed message in a single variable to cut down on duplication and improve readability (which I needed to chop away at to understand what was going on with HIP-RA code).

**Comment by malcolm-dsider** [comment]
The idea behind both printing and the logging the messages is that these different approaches serve two different classes of users – users who are running the Python app on their desktop via the command line need to see the messages printed to the screen; users who run the web app need to see error messages provided by the server, most likely generated in the structured log file. And server admins are used to looking into the server log files to track down errors.

 

If there was a way to tell the logging software to do both things, perhaps via a switch, then I’d be happy to drop the print statements. Or perhaps we build our own logging module that handles print, logging (using the same logging library we are already using), and/or both with just a one-line call in our code?

**Comment by softwareengineerprogrammer** [comment]
> If there was a way to tell the logging software to do both things, perhaps via a switch, then I’d be happy to drop the print statements.

Yes this is possible by configuring multiple handlers for the logger. In fact, this seems to already be present in `logging.conf` - the `consoleHandler` is configured as a `StreamHandler` with `sys.stdout`: https://github.com/NREL/GEOPHIRES-X/blob/bcaaf2120b591821a23649d68776ec84938a2212/src/geophires_x/logging.conf#L14-L18

I have looked at the logging configuration with some amount of scrutiny previously but deferred any major changes as it's always been comparatively low priority. I don't recall the exact conclusions, if any, I came to about what was working and not working and what should be changed, but here are a few key points that occur to me with a fresh look just now:

Having just looked now, I think the first thing I'd do is change line 16 from `WARN` to `INFO` as setting `WARN` level means `INFO` messages will be suppressed. I suspect the duplicitous use of `print` arose because someone tried `logger.info`, didn't see anything printed to the console because `INFO` is suppressed by the current settings, and thus they added `print` to brute-force output to the console instead of changing the handler level.

Another issue is the passing of the model object to its composite children classes for the sole purpose of providing access to the logger; example: https://github.com/NREL/GEOPHIRES-X/blob/8533d39ea7d1ec537f900435abc4a1c37445aeee/src/geophires_x/LHSReservoir.py#L44-L55

This creates a web of unnecessary dependency which causes a number of issues with typing and class inheritance (IIRC one of these issues was I tried adding type annotations to methods that take the model but was unable to do so because it created a circular dependency error). I *think* the correct solution to this is to simply add a private logger field to each class instead of passing the model, as I did in this PR in `GeoPHIRESUtils`:

https://github.com/NREL/GEOPHIRES-X/blob/8cbcc999b6946c9871ecb9d2fe4b00714334f7dd/src/geophires_x/GeoPHIRESUtils.py#L17

---

So to summarize:
1. I'll just re-iterate that the specific change I made here was not a change to behavior in logging nor did it have to do with logging behavior specifically; it was just basic programming best practice of declaring a variable instead of repeated hardcoded constants where the hardcoded constants in question just happened to be messages to be logged.
1. I would continue to treat proactive refactoring/overhaul of logging as low-priority for now - if it ain't broke, don't fix it
1. I created tracking issues capturing the enhancements that could be made when bandwidth/priority permit:
    1.  https://github.com/NREL/GEOPHIRES-X/issues/111 
    2. https://github.com/NREL/GEOPHIRES-X/issues/112
1. Follow the private logger pattern as in https://github.com/NREL/GEOPHIRES-X/blob/8cbcc999b6946c9871ecb9d2fe4b00714334f7dd/src/geophires_x/GeoPHIRESUtils.py#L17 for any required new instances of loggers

**Pull request merged**